python/lsst/ip/isr/deferredCharge.py

```diff
@@ -1067,33 +1086,45 @@ def run(self, exposure, ctiCalib, gains=None):
         gains : `dict` [`str`, `float`]
             A dictionary, keyed by amplifier name, of the gains to
             use.  If gains is None, the nominal gains in the amplifier
-            object are used.
+            object are used if necessary.
 
         Returns
         -------
         exposure : `lsst.afw.image.Exposure`
             The corrected exposure.
+
+        Notes
+        -------
+        This task will read the exposure metadata and determine if
+        applying gains if necessary. The correction takes place in
+        units of electrons. If bootstrapping, the gains used
+        will just be 1.0. and the input/output units will stay in
+        adu. If the input image is in adu, the output image will be
+        in units of electron. If the input image is in electron, the
+        output image will be in electron.
         """
         image = exposure.getMaskedImage().image
         detector = exposure.getDetector()
 
-        # If gains were supplied, they should be used.  If useGains is
-        # true, but no external gains were supplied, use the nominal
-        # gains listed in the detector.  Finally, if useGains is
-        # false, fake a dictionary of unit gains for ``gainContext``.
-        useGains = True
-        if "USEGAINS" in ctiCalib.getMetadata().keys():
-            useGains = ctiCalib.getMetadata()["USEGAINS"]
-            self.log.info(f"useGains = {useGains} from calibration metadata.")
-        else:
-            useGains = self.config.useGains
-            self.log.info(f"USEGAINS not found in calibration metadata.  Using {useGains} from config.")
+        # Get the image and overscan units.
+        imageUnits = exposure.getMetadata().get("LSST ISR UNITS")
+
+        # The deferred charge correction assumes that everything is in
+        # electron units. Make it so:
+        applyGains = False
+        if imageUnits == "adu":
+            applyGains = True
 
-        if useGains:
+        # If we need to convert the image to electrons, check that gains
+        # were supplied, they should be used. If no external gains were
+        # supplied, use the nominal gains listed in the detector.
```

Comment:
I worry about this; these gains are *never* correct.  Is this ever a mode we would be running in?

Reply:
It now raises an error if no gains are supplied.